### PR TITLE
lint: integrate with external `cmake-lint`

### PR DESCRIPTION
**Pull request by yangyingchao**
When `enable_external_cmake_lint` is turned on and `cmake-lint` is installed, use it to provide linting.  The result from `cmake-lint` is merged with the built-in linting result.

screenshot-2024-06-25-10-22-56

### CODE REVIEW

File: src/gammar.rs [hunk 1]
```diff
@@ -15,8 +18,68 @@ pub struct ErrorInfo {
...
+) -> Option<ErrorInfo> {
+    let future_cmake_lint = if whole {
+        Some(run_cmake_lint(local_path))
+    } else {
+        None
+    };
+
+    let mut result = checkerror_inner(local_path, &source.lines().collect(), input);
+    if let Some(future) = future_cmake_lint {
+        if let Some(v) = block_on(future) {
```
Comment (Decodetalkers) [inline src/gammar.rs:35]: This should not use block_on, you can turn the function to async function

File: src/gammar.rs [hunk 2]
```diff
@@ -15,8 +18,68 @@ pub struct ErrorInfo {
     )>,
 }
 
-pub fn checkerror(local_path: &Path, source: &str, input: tree_sitter::Node) -> Option<ErrorInfo> {
-    checkerror_inner(local_path, &source.lines().collect(), input)
+pub fn checkerror(
```
Comment (Decodetalkers) [inline src/gammar.rs:21]: Make this one async fn,

File: src/config.rs [hunk 3]
```diff
@@ -47,19 +48,46 @@ impl Default for LintSuggestion {
...
+pub static CMAKE_LINT: Lazy<LintSuggestion> =
+    Lazy::new(|| CMAKE_LINT_CONFIG.command_upcase.clone().into());
+
+#[cfg(test)]
+mod tests {
+    use crate::config::CMAKE_LINT_CONFIG;
+
+    #[test]
+    fn tst_lint_config() {
+        println!("command_upcase: {}", (*CMAKE_LINT_CONFIG).command_upcase);
```
Comment (Decodetalkers) [inline src/config.rs:87]: I think unit test should use asset, println is useless

**Comment by Decodetalkers** [comment]
What is whole means? Can you add some comments? Thanks

**Comment by yangyingchao** [comment]
I'm not sure what a proper name would be, maybe you can provide me with some suggestions.

The name is used to disable the external linter when the server receives the `did_change` event. External linting is only applied after the file is opened or saved.

By the way, I'm not very familiar with Rust's async programming. It would be great if you could help me review the changes...

-- *Yang Yingchao*

File: src/gammar.rs [hunk 4]
```diff
@@ -15,8 +17,66 @@ pub struct ErrorInfo {
...
+        let error_info = result.get_or_insert(ErrorInfo { inner: vec![] });
+        for item in v.inner {
+            error_info.inner.push(item);
+        }
+    };
+
+    result
+}
+
+async fn run_cmake_lint(path: &Path) -> Option<ErrorInfo> {
```
Comment (Decodetalkers) [inline src/gammar.rs:43]: But.. seems this function does not have await
. Emmm

**Comment by Decodetalkers** [comment]
by the way, what does whole means?

**Comment by yangyingchao** [comment]
> by the way, what does whole means?

I'm not sure what a proper name would be, maybe you can provide me with some suggestions.

The name is used to disable the external linter when the server receives the `did_change` event. External linting is only applied after the file is opened or saved.

By the way, I'm not very familiar with Rust's async programming. It would be great if you could help me review the changes...

**Comment by Decodetalkers** [comment]
ha.. I see

**Comment by Decodetalkers** [comment]
thanks for your pr!

File: src/gammar.rs (hunk 4 above)
Comment (yangyingchao) [reply src/gammar.rs]: I made this function async because I want to run external linter in parallel with the built-in one..

**Comment by Decodetalkers** [comment]
but... https://github.com/richq/cmake-lint this one is archive.. which repo cmake-lint is in?

**Comment by yangyingchao** [comment]
> but... https://github.com/richq/cmake-lint this one is archive.. which repo cmake-lint is in?
> 
> oh,, https://github.com/cmake-lint/cmake-lint it is this repo, it is actively maintained.


Actually, I'm using the old one, which is now part of cmakelang which consists of cmake-format & cmake-lint and other tools:
https://cmake-format.readthedocs.io/en/latest/index.html

It is not active indeed, but usable...

If you prefer to cmakelint  (https://github.com/cmake-lint/cmake-lint) , the regex part needs to be updated accordingly.

**Comment by Decodetalkers** [comment]
> > but... https://github.com/richq/cmake-lint this one is archive.. which repo cmake-lint is in?
> > oh,, https://github.com/cmake-lint/cmake-lint it is this repo, it is actively maintained.
> 
> Actually, I'm using the old one, which is now part of cmakelang which consists of cmake-format & cmake-lint and other tools: https://cmake-format.readthedocs.io/en/latest/index.html
> 
> It is not active indeed, but usable...
> 
> If you prefer to cmakelint (https://github.com/cmake-lint/cmake-lint) , the regex part needs to be updated accordingly.

I think we should use the actively maintained one

**Comment by yangyingchao** [comment]
> > > but... https://github.com/richq/cmake-lint this one is archive.. which repo cmake-lint is in?
> > > oh,, https://github.com/cmake-lint/cmake-lint it is this repo, it is actively maintained.
> > 
> > 
> > Actually, I'm using the old one, which is now part of cmakelang which consists of cmake-format & cmake-lint and other tools: https://cmake-format.readthedocs.io/en/latest/index.html
> > It is not active indeed, but usable...
> > If you prefer to cmakelint (https://github.com/cmake-lint/cmake-lint) , the regex part needs to be updated accordingly.
> 
> I think we should use the actively maintained one

To be clear: `cmake-lint`, which is integrated in this PR, is a tool within cmakelang. It has nothing to do with the two projects you mentioned.

And I think it is better because it has different levels for the lints, just like shellcheck does.

**Comment by Decodetalkers** [comment]
> > > > but... https://github.com/richq/cmake-lint this one is archive.. which repo cmake-lint is in?
> > > > oh,, https://github.com/cmake-lint/cmake-lint it is this repo, it is actively maintained.
> > > 
> > > 
> > > Actually, I'm using the old one, which is now part of cmakelang which consists of cmake-format & cmake-lint and other tools: https://cmake-format.readthedocs.io/en/latest/index.html
> > > It is not active indeed, but usable...
> > > If you prefer to cmakelint (https://github.com/cmake-lint/cmake-lint) , the regex part needs to be updated accordingly.
> > 
> > 
> > I think we should use the actively maintained one
> 
> To be clear: `cmake-lint`, which is integrated in this PR, is a tool within cmakelang. It has nothing to do with the two projects you mentioned.
> 
> And I think it is better because it has different levels for the lints, just like shellcheck does.

I see

**Pull request merged**